middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -428,49 +428,64 @@ let translate_apply0 ~dbg_with_inlined:dbg env res apply =
       in
       C.resume ~dbg ~stack ~f ~arg ~last_fiber, free_vars, env, res, Ece.all)
 
-(* Function calls that have an exn continuation with extra arguments must be
-   wrapped with assignments for the mutable variables used to pass the extra
-   arguments. *)
-(* CR mshinwell: Add first-class support in Cmm for the concept of an exception
-   handler with extra arguments. *)
 let translate_apply env res apply =
   let dbg = Env.add_inlined_debuginfo env (Apply.dbg apply) in
   warn_if_unused_inlined_attribute apply ~dbg_with_inlined:dbg;
   let call, free_vars, env, res, effs =
     translate_apply0 ~dbg_with_inlined:dbg env res apply
   in
   let k_exn = Apply.exn_continuation apply in
-  let mut_vars =
-    Exn_continuation.exn_handler k_exn |> Env.get_exn_extra_args env
-  in
   let extra_args = Exn_continuation.extra_args k_exn in
-  if List.compare_lengths extra_args mut_vars = 0
-  then
-    (* Note wrt evaluation order: this is correct for the same reason as
-       `To_cmm_shared.simple_list`, namely the first simple translated (and
-       potentially inlined/substituted) is evaluted last. *)
-    let aux (call, env, res, free_vars) (arg, _k) v =
-      let To_cmm_env.
-            { env;
-              res;
-              expr = { cmm = arg; free_vars = arg_free_vars; effs = _ }
-            } =
-        C.simple ~dbg env res arg
-      in
-      let free_vars = Backend_var.Set.union free_vars arg_free_vars in
-      C.sequence (C.assign v arg) call, env, res, free_vars
+  if Misc.Stdlib.List.is_empty extra_args
+  then call, free_vars, env, res, effs
+  else
+    (* If there are extra args, create a wrapper continuation, as we do when
+       inlining a function into a context where the exception continuation takes
+       extra args. *)
+    let env, res, free_vars, effs, extra_args_rev =
+      (* Obtain the Cmm expressions for the extra args *)
+      List.fold_left
+        (fun (env, res, free_vars, effs, extra_args_rev)
+             (extra_arg_simple, _kind) ->
+          let { To_cmm_env.env;
+                res;
+                expr = { cmm = extra_arg; free_vars = free_vars'; effs = effs' }
+              } =
+            C.simple ~dbg env res extra_arg_simple
+          in
+          ( env,
+            res,
+            Backend_var.Set.union free_vars free_vars',
+            Ece.join effs effs',
+            extra_arg :: extra_args_rev ))
+        (env, res, free_vars, effs, [])
+        extra_args
     in
-    let call, env, res, free_vars =
-      List.fold_left2 aux (call, env, res, free_vars) extra_args mut_vars
+    let extra_args = List.rev extra_args_rev in
+    let exn_var = Backend_var.create_local "*exn*" in
+    let handler_cont = Lambda.next_raise_count () in
+    let push_cont = Lambda.next_raise_count () in
+    let body =
+      C.create_ccatch ~rec_flag:false
+        ~body:(C.cexit push_cont [] [Push handler_cont])
```

Comment:
I think there's a bug here: this `Push` doesn't have a corresponding `Pop`. I think you will have to add another wrapper continuation, for the normal return continuation, that pops the trap properly.